### PR TITLE
Hydrator++ UI bug fix

### PR DESCRIPTION
**Pull request by ajainarayanan**
- CDAP-5619, CDAP-5557: Show appropriate message in bottom panel when imported config json doesn't have artifact information or has invalid artifact information.
- CDAP-5297: Handling pipelines created using old hydrator. Show warning message for pipelines created using old hydrator and prevents showing detailed view requesting user to upgrade their pipeline.
- CDAP-4692: Prevents user from accidentally navigating away from the studio view. Shows browser native alert message before proceeding with the navigation.
- Fixes the trash icon in the node config panel not overflow its container.

### CODE REVIEW

File: cdap-ui/app/features/hydratorplusplus/services/hydrator-service.js [hunk 1]
```diff
@@ -122,16 +122,27 @@ class HydratorPlusPlusHydratorService {
...
       pluginName: node.plugin.name
     };
-    let nodeArtifact = node.plugin.artifact;
+
     return this.myPipelineApi.fetchPluginProperties(params)
       .$promise
       .then(function(res = []) {
+        let nodeArtifact = node.plugin.artifact;
+        if (node.plugin && !node.plugin.artifact && res.length) {
+          let lastElement = res.length - 1;
```
Comment (anwar6953) [inline cdap-ui/app/features/hydratorplusplus/services/hydrator-service.js]: nit - this is more like `lastIndex`.

File: cdap-ui/app/features/hydratorplusplus/routes.js [hunk 2]
```diff
@@ -138,13 +138,45 @@ angular.module(PKG.name + '.feature.hydratorplusplus')
...
+                      $state.go('hydrator.list');
+                      return;
+                    }
+                    if(!config.stages) {
+                      myAlertOnValium.show({
+                        type: 'danger',
+                        content: 'Pipeline is created using older version of hydrator. Please upgrage the pipeline to newer version(3.4) to view in UI.'
+                      });
+                      $q.reject(false);
+                      // FIXME: We should not have done this. But ui-router when rejected on a 'resolve:' function takes it to the parent state apparaently
```
Comment (elfenheart) [inline cdap-ui/app/features/hydratorplusplus/routes.js]: nit: typo ```apparently```

File: cdap-ui/app/features/hydratorplusplus/services/hydrator-service.js [hunk 3]
```diff
@@ -122,16 +122,27 @@ class HydratorPlusPlusHydratorService {
...
+          node._backendProperties = res[lastElement].properties || {};
+          node.description = res[lastElement].description;
+          node.plugin.artifact = res[lastElement].artifact;
+          defer.resolve(node);
+          return defer.promise;
```
Comment (elfenheart) [inline cdap-ui/app/features/hydratorplusplus/services/hydrator-service.js]: shouldn't this return statement be outside the ```if``` block?

File: cdap-ui/app/features/hydratorplusplus/services/hydrator-service.js [hunk 4]
```diff
@@ -122,16 +122,27 @@ class HydratorPlusPlusHydratorService {
...
+        }
```
Comment (elfenheart) [inline cdap-ui/app/features/hydratorplusplus/services/hydrator-service.js]: I think we should only have 1 ```return defer.promise``` in this callback and just resolve different things based on the condition. Seeing 2 defer.promise was a little weird

**Comment by elfenheart** [comment]
Couple of issue:
- After making some changes, clicking on Studio link at navbar will trigger the confirmation box, but not actually doing anything
- The alert when artifact information is incorrect does not show when the entire artifact object in JSON config is not present at all.

**Comment by anwar6953** [comment]
no further comments from me; lgtm

**Comment by elfenheart** [comment]
👍  LGTM

**Pull request merged**